src/jon_mikkel_ex/ex06/logistic_regression.py

```diff
@@ -0,0 +1,333 @@
+            The step-size for the gradient descent updates.
+        random_state : np.random.random_state or int or None (default=None)
+        A numpy random state object or a seed for a numpy random state object.
+        """
+        self.max_iter = max_iter
+        self.tol = tol
+        self.learning_rate = learning_rate
+        self.random_state = random_state
+
+    def _has_converged(self, coef, X, y):
+        r"""Whether the gradient descent algorithm has converged.
+
+        Returns True if the norm of the gradient is smaller than ``self.tol``,
+        mathematically, that is
+
+        .. math::
+
+            ||\nabla_w L(\mathbf{w}^{(k)}; X, \mathbf{y})|| < T
+
+        where :math:`\nabla_w L` is the gradient of the loss function,
+        :math:`|| \mathbf{v} ||` is the norm of the vector :math:`\mathbf{v}`,
+        :math:`\mathbf{w}^{(k)}` is the weights at iteration ``k``, and
+        :math:`T` is the convergence tolerance (``self.tol``).
+
+        Parameters
+        ----------
+        coef : np.ndarray(shape=(r,))
+            The weight vector, :math:`\mathbf{w}^{(k)}`
+        X : np.ndarray(shape=(n, r))
+            The data matrix (aka design or measurement matrix)
+        y : np.ndarray(shape=(n,))
+            The true class labels for each data point.
+
+        Returns
+        -------
+        has_converged : bool
+            True if the convergence criteria above is met, False otherwise.
+        """
+        return norm(logistic_gradient(coef, X, y)) < self.tol
+
+    def _fit_gradient_descent(self, coef, X, y):
+        r"""Fit the logisitc regression model to the data given initial weights
+
+        Gradient descent works by iteratively applying the following update
+        rule
+
+
+
+        where :math:`\mathbf{w}^{(k)}` is the coefficient vector at iteration 
+        ``k``, :math:`\mathbf{w}^{(k-1)}` is the coefficient vector at 
+        iteration k-1, :math:`\eta` is the learning rate and 
+        :math:`\nabla L(\mathbf{w}^{(k-1)}; X, \mathbf{y})` is the gradient of
+        the loss function at iteration k-1.
+
+        The iterative algorithm should be performed for at most
+        ``self.max_iter`` iterations, or until the convergence criteria is
+        reached.
+
+        Parameters
+        ----------
+        coef : np.ndarray(shape=(r,))
+            The initial guess for the coefficient vector.
+            May be modified inplace by the method.
+        X : np.ndarray(shape=(n, r))
+            The data matrix
+        y : np.ndarray(shape=(n,))
+            The target vector
+
+        Returns
+        -------
+        coef : np.ndarray(shape=(n,))
+            The logistic regression weights
+        """
+        for _ in range(self.max_iter):
+            coef += (self.learning_rate * logistic_gradient(coef, X, y))
+            if self._has_converged(coef, X, y):
+                break
+        return coef  # Returns the first coefficient that converges enough
```

Comment:
I would recommend using a while loop here with a double control statement:
```python
while _iter <= learning rate and not _has_converged():
iter += 1
...
```

Reply:
If the function would be longer, its easier to read if the exit statement is in the beginning rather than having to look for a `break` somewhere in the function.